proposals/stdlib/mutablelist-move.md

```diff
@@ -0,0 +1,145 @@
+# Adding moveXXX() Functions to MutableList
+
+It would be helpful to create some additional extension functions for `MutableList` to move items to new indexes, as well as move them up/down, and move one or all items that meet a predicate.
+
+
+# Title
+
+* **Type**: Standard Library API proposal
+* **Author**: Thomas Nield
+* **Contributors**: N/A
+* **Status**: Submitted
+* **Prototype**: Implemented
+
+
+## Summary
+
+It would be helpful to create some additional extension functions for `MutableList` to move items to new indexes, as well as move them up/down, and one or all items that meet a predicate.
+
+These can be implemented using only extension functions on `MutableList`.
+
+## Similar API review
+
+Cannot think of an explicit precendent in another API.
+
+## Use cases
+
+* Common use cases often have to do with re-ordering items especially in a UI context, and any scenario dealing with mutable ordered collections may be candidate.
+
+* In JavaFX and other UI technologieis, a user may want to move an item in a `MutableList` up or down to define some arbitrary order.
+
+* Items can be searched in a `MutableList` by having the matching items moved to the top of the `MutableList`, rather than creating a new `List` or some filtering abstraction
+
+* Observe this [demo application](https://github.com/thomasnield/rxkotlinfx-tornadofx-demo) to see several of these use cases in action.
+
+## Alternatives
+
+Developers could implement the extension functions themselves, as shown below. Moving items can be tricky and verbose to coordinate with local implementations.
+
+```kotlin
+/**
+ * Moves the given **T** item to the specified index
+ */
+fun <T> MutableList<T>.move(item: T, newIndex: Int)  {
+    val currentIndex = indexOf(item)
+    if (currentIndex < 0) return
+    removeAt(currentIndex)
+    add(newIndex, item)
+}
+
+/**
+ * Moves the given item at the `oldIndex` to the `newIndex`
+ */
+fun <T> MutableList<T>.move(oldIndex: Int, newIndex: Int)  {
+    val item = this[oldIndex]
+    removeAt(oldIndex)
+    add(newIndex, item)
+}
+
+/**
+ * Moves all items meeting a predicate to the given index
+ */
+fun <T> MutableList<T>.moveAll(newIndex: Int, predicate: (T) -> Boolean) {
+
+    var newIndexIncrement = newIndex
+
+    asSequence().toList().asSequence().withIndex()
+            .filter { v -> predicate.invoke(v.value) }
+            .forEach { move(it.index, newIndexIncrement++) }
```

Comment:
Imagine you have items with indices 0 and 1 matching the predicate and you want to move them to the end. First you move an item with index 0 to the end and all other items are moved to the beginning. Then you try to move an item with index 1, but it's no more that item that was matched by predicate.